azure-iot-device/azure/iot/device/provisioning/internal/polling_machine.py

```diff
@@ -0,0 +1,325 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+import logging
+import uuid
+from threading import Timer
+from transitions import Machine
+from ..transport import constant
+
+from .action import SubscribeAction, PublishAction, UnsubscribeAction
+from .request_response_provider import RequestResponseMachine
+
+
+logger = logging.getLogger(__name__)
+
+
+class PollingMachine(object):
+    """
+    Class that is responsible for sending the initial registration request and polling the
+    registration process for constant updates.
+    """
+
+    def __init__(self, state_based_provider):
+        """
+        :param state_based_provider: The state machine based provider.
+        """
+        self._polling_timer = None
+        self._query_timer = None
+
+        self._register_callback = None
+        self._cancel_callback = None
+
+        self.on_connected = None
+        self.on_disconnected = None
+        self.on_registration_status_update = None
+
+        self._operations = {}
+
+        self._request_response_provider = RequestResponseMachine(state_based_provider)
+
+        self._request_response_provider.on_response_received = self._handle_response_received
+
+        states = [
+            "disconnected",
+            "initializing",
+            "registering",
+            "waiting_to_poll",
+            "polling",
+            "completed",
+            "error",
+            "cancelling",
+        ]
+
+        transitions = [
+            {
+                "trigger": "_trig_register",
+                "source": "disconnected",
+                "before": "_initialize_register",
+                "dest": "initializing",
+            },
```

Comment:
I'm not sure how this works.  If you're in the disconnected state, and you call register, then _trig_register gets called.  It starts to connect and goes into the intializing state and then I think it get stuck.  It looks like you have to call register a second time before it sends the registration request.

Reply:
the way i did this is while initializing (during `_initialize_register`) i am connecting as well as send a subscribe request as a post connect action. When the SUBACK comes back another trigger is fired which takes it from initializing to registering. This actually publishes the registration request.

---
In reply to: [277410947](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r277410947) [](ancestors = 277410947)

Reply:
Ah, OK, I did notice that after I wrote this comment.  That seems confusing.  _trig_register does two things then.  It either subscribes or it registers.  I'm not saying that your code doesn't work, I'm just saying that it seems like you're overloading the trigger -- sometimes it it initiated by the user, sometimes it is initiated as a result of a suback.  sometimes it causes a subscription, sometimes it causes a publish.  

---
In reply to: [277466679](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r277466679) [](ancestors = 277466679,277410947)

Reply:
Oops.  I didn't see _trig_send_registration_request.  Ignore my last comment.

---
In reply to: [277790233](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r277790233) [](ancestors = 277790233,277466679,277410947)

Reply:
I agree with Bert

---
In reply to: [277790233](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r277790233) [](ancestors = 277790233,277466679,277410947)